### PR TITLE
Fixes #6710 - unicode characters in url parameters

### PR DESCRIPTION
**Pull request by tstrachota**
Original methods to_param defined on resources called name.parameterize
to get rid of url-unsafe characters. This function also strips off unicode
characters.
Another problem was in one unified search function for all parameterization approaches. It used to fail when a name was parameterized to blank string or when a name contained only numbers.

Changes:
- parameterization extracted into a separate module Parameterizable
- each resource defines methods 'to_param' and 'from_param' to ensure correct search
- views updated to always use to_param when building links
- parameterization changed to id+name for resources where it wasn't safe to rely only on names
- default parameter is id

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -121,18 +121,26 @@ def model_of_controller
...
-  # searches for an object based on its name and assign it to an instance variable
-  # required for models which implement the to_param method
+  # Searches for a record based on it's parameterized version. If the model
+  # does not respond to 'from_param', search by name or id is used to keep backward
+  # compatibility.
   #
   # example:
-  # @host = Host.find_by_name params[:id]
+  # before_filter :find_by_name, :only => %w{edit update destroy}
   def find_by_name
```
Comment (tstrachota) [inline app/controllers/application_controller.rb]: I kept the original method name for backwards compatibility with plugins but it doesn't make much sense now. Should I rename it, what do you think?

File: app/models/concerns/parameterizable.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    end
+  end
+
+  module ByIdName
+    extend ActiveSupport::Concern
+
+    included do
+      def to_param
+        # remove characters unsafe for urls, keep unicode ones
+        "#{id}-#{name}".gsub(/[!*'();:@&=+$,\/?%#\[\]]/, '').chomp('-')
```
Comment (ohadlevy) [inline app/models/concerns/parameterizable.rb]: did you consider using http://apidock.com/rails/ActiveSupport/Inflector/transliterate ?

File: app/models/fact_name.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 class FactName < ActiveRecord::Base
+  include Parameterizable::ByIdName
```
Comment (ohadlevy) [inline app/models/fact_name.rb:2]: byname or by idname ?

File: app/models/lookup_key.rb [hunk 4]
```diff
@@ -1,5 +1,6 @@
 class LookupKey < ActiveRecord::Base
   include Authorizable
+  include Parameterizable::ById
```
Comment (ohadlevy) [inline app/models/lookup_key.rb]: i wonder if thats not going to be the default regardless ? e.g. do we need to explicitly define that in every model?

File: config/initializers/active_record_extensions.rb [hunk 5]
```diff
@@ -2,4 +2,5 @@ class ActiveRecord::Base
   extend Host::Hostmix
   include HasManyCommon
   include StripWhitespace
-end
\ No newline at end of file
+  include Parameterizable::ById
```
Comment (ohadlevy) [inline config/initializers/active_record_extensions.rb:5]: if this is included, I assume we dont need it in other models (such as lookupkey) ?

Comment (tstrachota) [inline config/initializers/active_record_extensions.rb:5]: Correct. That line in lookupkeys was superfluous.

File: app/models/concerns/parameterizable.rb (hunk 2 above)
Comment (tstrachota) [inline app/models/concerns/parameterizable.rb]: I did but `transliterate` replaces uknown characters with `?` which could result in having plenty of question marks instead of names (e.g. in Chinese).
From the tests I did, Rails encodes unicode characters in urls well with %. Browser translates the encoded characters back in the address bar so there's no loss of readability.

File: app/models/fact_name.rb (hunk 3 above)
Comment (tstrachota) [inline app/models/fact_name.rb:2]: I intentionally changed it to id-name because I didn't find any format validation for `FactName` and considered it not safe for usage in urls.
Is the name format guaranteed by any other conditions (e.g. puppet restrictions)?

**Comment by GregSutcliffe** [comment]
Needs rebasing

**Comment by mbacovsky** [comment]
Rebased

**Comment by ohadlevy** [comment]
needs another rebase it seems :)

**Comment by mbacovsky** [comment]
Rebased

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -121,18 +121,26 @@ def model_of_controller
     @model_of_controller ||= controller_path.singularize.camelize.gsub('/','::').constantize
   end
 
-
-  # searches for an object based on its name and assign it to an instance variable
-  # required for models which implement the to_param method
+  # Searches for a record based on it's parameterized version. If the model
```
Comment (dLobatog) [inline app/controllers/application_controller.rb]: `it's` -> `its`

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (dLobatog) [inline app/controllers/application_controller.rb]: Probably now you should use `friendly_id` after this commit was merged. https://github.com/theforeman/foreman/commit/8b737c9c7648b3726dadb3b2e4708fcb43af02a8

**Comment by dLobatog** [comment]
This needs rebasing now that we use `friendly_id`, I expect some changes will be necessary.

**Comment by dLobatog** [comment]
Note from discussion on IRC - NestedAncestryCommon is overriding to_param for taxonomies and host groups at least, so that a root host group with all unicode chars will become ``#{id}-`. I guess we should use the unicode chars. This might be happening at other places since `Parametrize::ById` is loaded very early compared to concerns or class methods.

**Comment by tstrachota** [comment]
Updated. I fixed some more occurrences of `to_param` that remained without attention after the rebases. All should be fine now.
Thanks for catching that @eLobato.

**Comment by dLobatog** [comment]
@tstrachota Awesome, could you just add one or two test to check unicode is 'translated' in the URL? For models with NestedAncestryCommon and models without it

**Comment by tstrachota** [comment]
@eLobato Updated. I added tests for `parameterize` and `to_param` on hostgroups.

**Comment by dLobatog** [comment]
@tstrachota There's a bug introduced, host groups and other stuff that uses ancestry, is now encoding the URLs without dashes to separate parent/children, like: 

`http://localhost:3000/hostgroups/1673-adcmonfrontierfrontdev/edit`

It should be 

`http://localhost:3000/hostgroups/1663-adcmon-frontier/edit`

It does it for non unicode stuff and unicode stuff, for example

`http://localhost:3000/hostgroups/1767-acron%D1%A0%D1%A1%D1%A2%D1%A3%D1%A4%D1%A5%D1%A6%D1%A7%D1%A8%D1%A9%D1%AA%D1%AB%D1%AC%D1%AD%D1%AE%D1%B7%D1%B8%D1%B9%D1%BA%D1%BB%D1%BC%D1%BD%D1%BE%D1%BF/edit`

Please fix and add a tests for these dashes and I think we're good to go :rocket:  :+1:

**Comment by tstrachota** [comment]
@eLobato Parameterization updated, changes reflected in tests:
https://github.com/theforeman/foreman/pull/1685/files#diff-a620d136f1e20e88384db07470f9c124R12
https://github.com/theforeman/foreman/pull/1685/files#diff-ec15c79bfabb56e6ef525d4f7c65808bR334

**Comment by dLobatog** [comment]
@tstrachota Thanks, it's looking fine. Merging in 3..2..1 :bomb:

**Comment by dLobatog** [comment]
Merged as https://github.com/theforeman/foreman/commit/e768c9761528c64fcc289c391091fcce63e82ff8, thanks @tstrachota
